1단계 - 레거시 코드 리팩터링

KATEKEITH opened the pull request:
힌트에 유의하여 구현했습니다 : )

KATEKEITH committed:
refactor: QnaService.deleteQuestion() 비즈니스 로직 도메인으로 이동

- 핵심 비즈니스 로직을 Question.delete()로 위임
- 서비스는 인프라 관련 작업만 담당
- 기존 QnaServiceTest 모든 테스트 통과 유지
- 단위 테스트 가능한 도메인 중심 설계로 개선

KATEKEITH committed:
test: Question.delete() 도메인 메서드 테스트 케이스 추가

비즈니스 로직이 도메인으로 이동한 후 단위 테스트 케이스 추가
- 외부 의존성 없는 순수 도메인 로직 테스트
- 질문 삭제 시나리오에 대한 테스트 커버리지 확보

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -85,6 +86,31 @@ public List<Answer> getAnswers() {
         return answers;
     }
 
+    public List<DeleteHistory> delete(NsUser loginUser) throws CannotDeleteException {
```
이 메서드에 대한 단위 테스트 추가할 수 있지 않을까?

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -85,6 +86,31 @@ public List<Answer> getAnswers() {
+            throw new CannotDeleteException("질문을 삭제할 권한이 없습니다.");
+        }
+
+        for (Answer answer : answers) {
+            if (!answer.isOwner(loginUser)) {
+                throw new CannotDeleteException("다른 사람이 쓴 답변이 있어 삭제할 수 없습니다.");
+            }
+        }
```
Answers라는 이름의 일급 콜렉션을 추가해 보면 어떨까?

javajigi commented on src/main/java/nextstep/qna/domain/Question.java:
```diff
@@ -85,6 +86,31 @@ public List<Answer> getAnswers() {
+        deleteHistories.add(new DeleteHistory(ContentType.QUESTION, this.getId(), this.writer, LocalDateTime.now()));
+
+
+        for (Answer answer: answers) {
+            answer.setDeleted(true);
+            deleteHistories.add(new DeleteHistory(ContentType.ANSWER, answer.getId(), answer.getWriter(), LocalDateTime.now()));
+        }
```
이와 같이 구현할 경우 Answer의 상태를 바꾸기 위해 setter 메서드를 사용해야 한다.
삭제 가능 여부를 판단하고 삭제 상태로의 변경을 Answer가 담당하도록 메시지를 보내보면 어떨까?

javajigi reviewed (changes requested):
첫 번재 리뷰 요청이 미션 중 가장 중요하다고 할 수 있는 마지막 미션 1단계 요청이네요. 👍 
포기하지 않고 한 번이라도 리뷰 요청한 부분 칭찬해요.
Service 레어이 로직은 도메인 객체로 이동한 점 👍 
단, Question만 사용하지 말고 Answer도 적극 사용해 보면 어떨까요?

KATEKEITH committed:
refactor: 삭제 상태로의 변경을 Answer가 담당하도록 메시지를 보낸다.

KATEKEITH committed:
feat: 강의 이미지 도메인 구현 및 유효성 검증 추가

- 이미지 크기, 타입, 최소 너비/높이, 3:2 비율 검증 로직 포함
- SessionImageType enum을 통해 확장 가능한 타입 관리
- 잘못된 입력에 대해 IllegalArgumentException 처리
